tests/transforms/test_transform_can_be_hoisted.py

```diff
@@ -0,0 +1,99 @@
+from xdsl.builder import ImplicitBuilder
+from xdsl.dialects import func, scf, test
+from xdsl.dialects.arith import Addf
+from xdsl.dialects.builtin import (
+    FunctionType,
+    IndexType,
+    ModuleOp,
+    f32,
+    i32,
+)
+from xdsl.dialects.scf import For, Yield
+from xdsl.ir import Block, Region
+from xdsl.transforms.loop_invariant_code_motion import can_be_hoisted
+
+index = IndexType()
+
+
+def test_for_with_loop_invariant_verify():
+    """Test for with loop-variant variables"""
+
+    lower = test.TestOp(result_types=[i32])
+    upper = test.TestOp(result_types=[i32])
+    step = test.TestOp(result_types=[i32])
+    carried = test.TestOp(result_types=[i32])
+
+    a = test.TestOp(result_types=[i32])
+    b = test.TestOp(result_types=[i32])
+    block = Block(
+        arg_types=(
+            index,
+            index,
+        )
+    )
+    region = Region(block)
+    with ImplicitBuilder(block) as (arg0, _):
+        op1 = test.TestOp(result_types=[i32])
+        op2 = test.TestOp(result_types=[i32])
+        op3 = test.TestOp([op1, op2, arg0], result_types=[i32])
+        f = Yield(op3)
+    for_op = For(lower, upper, step, [carried], region)
+    block0 = Block([a, b, for_op])
+    _region_outer = Region(block0)
+
+    assert can_be_hoisted(op1, region)
+    assert can_be_hoisted(op2, region)
+    assert not can_be_hoisted(op3, region)
+    assert not can_be_hoisted(f, region)
+
+
+def test_invariant_loop_dialect():
+    """Test for with loop-variant variables"""
+    # Create constants
+
+    ci0 = test.TestOp(result_types=[i32])
+    cf7 = test.TestOp(result_types=[f32])
+    cf8 = test.TestOp(result_types=[f32])
+    ci10 = test.TestOp(result_types=[i32])
+    ci1 = test.TestOp(result_types=[i32])
+    co0 = test.TestOp(result_types=[i32])
+    co1 = test.TestOp(result_types=[i32])
+    coi = test.TestOp(result_types=[i32])
+    block = Block(
+        arg_types=(
+            index,
+            index,
+        )
+    )
+    block1 = Block(
+        arg_types=(
+            index,
+            index,
+        )
+    )
+    region_inner = Region(block)
+    with ImplicitBuilder(block) as (_arg0, _):
+        op1 = test.TestOp(result_types=[i32])
+        # Test operation inside the loop
+        hello = test.TestOp([op1], result_types=[i32])
+        test.TestOp([hello], result_types=[])
+        # Floating-point addition inside the loop
+        v0 = Addf(cf7, cf8)
+
+    region = Region(block1)
+    with ImplicitBuilder(block1) as (_arg1, _):
+        scf.For(ci0, ci10, ci1, [], region_inner)
+
+    for_op = For(co0, co1, coi, [], region)
+    body0 = Block([cf7, cf8, for_op])
+    region_outer = Region(body0)
+    func_type = FunctionType.from_lists([], [])
+    function = func.FuncOp("invariant_loop_dialect", func_type, region_outer)
```

Comment:
This is now looking great, but the structure of the module is a little hard to read. The `ImplicitBuilder` regions can be nested, which will make it a little easier to figure out what's going on. Here is what I would suggest to do:

```suggestion
    outer_region = Region([Block()])
    with ImplicitBuilder(outer_region):
        ci32, cf32 = test.TestOp(result_types=[i32, f32]).results
        
        outer_for = For(ci32, ci32, ci32, [], Region(Block(arg_types=[index, index])
        with ImplicitBuilder(outer_for.body) as (_outer_i, _):
            inner_for = For(ci32, ci32, ci32, [], Region(Block(arg_types=[index, index])
            with ImplicitBuilder(inner_for.body) as (_inner_i, _):
                no_args = test.TestOp(result_types=[i32])
                args_in_inner = test.TestOp([no_args], result_types=[i32])
                no_results = test.TestOp([args_in_inner], result_types=[])
```

I removed the arith operation to avoid depending on that dialect, which is not relevant to this test. With this refactor, it is clear that the nested region functionality is not tested. It would be great to add another two testop in the loop, one of which has a region that is self-contained, the other of which depends on something outside of the for loop. I'm also not sure that the nested for loops get us anything.

Reply:
In general, could you please check that each of the conditions in the `can_be_hoisted` function are tested? I'm not 100% sure that this is the last missing example.

Reply:
I have made the changes